ambry-api/src/main/java/com/github/ambry/quota/QuotaChargeCallback.java

```diff
@@ -51,22 +50,23 @@ public void charge(long chunkSize) throws RouterException {
           if (throttlingRecommendation != null && throttlingRecommendation.shouldThrottle() && shouldThrottle) {
             if (quotaManager.getQuotaMode() == QuotaMode.THROTTLING
                 && quotaManager.getQuotaConfig().throttleInProgressRequests) {
-              throw new RouterException("RequestQuotaExceeded", RouterErrorCode.TooManyRequests);
+              throw new QuotaException("Exception while charging quota",
+                  new RouterException("RequestQuotaExceeded", RouterErrorCode.TooManyRequests), false);
             } else {
               logger.debug("Quota exceeded for an in progress request.");
             }
           }
         } catch (Exception ex) {
-          if (ex instanceof RouterException && ((RouterException) ex).getErrorCode()
-              .equals(RouterErrorCode.TooManyRequests)) {
+          if (ex.getCause() instanceof RouterException && ((RouterException) ex.getCause()).getErrorCode()
+              .equals(RouterErrorCode.TooManyRequests) && quotaManager.getQuotaMode() == QuotaMode.THROTTLING) {
```

Comment:
seems like if we have router error code being TooManyRequests, the quotaMode should already be THROTTLING, this is redundant.